tars/util/gpool/gpool.go

```diff
@@ -0,0 +1,86 @@
+package gpool
+
+type Worker struct {
+	WorkerQueue chan *Worker
+	JobChannel  chan Job
+	Stop        chan struct{}
+}
+
+func (w *Worker) Start() {
+	go func() {
+		var job Job
+		for {
+			w.WorkerQueue <- w
+			select {
+			case job = <-w.JobChannel:
+				job()
+			case <-w.Stop:
+				w.Stop <- struct{}{}
+				return
+			}
+		}
+	}()
+}
+
+func newWorker(pool chan *Worker) *Worker {
+	return &Worker{
+		WorkerQueue: pool,
+		JobChannel:  make(chan Job),
+		Stop:        make(chan struct{}),
+	}
+}
+
+type Job func()
+
+type Pool struct {
+	JobQueue    chan Job
+	WorkerQueue chan *Worker
+	stop        chan struct{}
+}
+
+func NewPool(numWorkers int, jobQueueLen int) *Pool {
+	jobQueue := make(chan Job, jobQueueLen)
+	workerQueue := make(chan *Worker, numWorkers)
+
+	pool := &Pool{
+		JobQueue:    jobQueue,
+		WorkerQueue: workerQueue,
+		stop:        make(chan struct{}),
+	}
+	pool.Start()
+	return pool
+}
+
+func (p *Pool) Start() {
+	for i := 0; i < cap(p.WorkerQueue); i++ {
+		worker := newWorker(p.WorkerQueue)
+		worker.Start()
+	}
+
+	go p.dispatch()
+}
+
+func (p *Pool) dispatch() {
+	for {
+		select {
+		case job := <-p.JobQueue:
+			worker := <-p.WorkerQueue
+			worker.JobChannel <- job
+		case <-p.stop:
+			for i := 0; i < cap(p.WorkerQueue); i++ {
+				worker := <-p.WorkerQueue
+
+				worker.Stop <- struct{}{}
+				<-worker.Stop
+			}
+
+			p.stop <- struct{}{}
+			return
+		}
+	}
+}
+
+func (p *Pool) Release() {
+	p.stop <- struct{}{}
+	<-p.stop
```

Comment:
Seems wired here. p.stop <- ,  then <-p.stop  , what if  <-p.stop  get executed before select stament?

Reply:
> Seems wired here. p.stop <- , then <-p.stop , what if <-p.stop get executed before select stament?

Please look at https://github.com/TarsCloud/TarsGo/blob/45a20271e0d699c24978e66818c9f1b4b746741e/tars/util/gpool/gpool.go#L69-L78

`p.stop <- struct{}{}` will send msg to above codes and wait for the workers to stop. When all workers are all stopped, it will also send `p.stop <- struct{}{}`.